### PR TITLE
#651 Adding missing unit tests to Inspector classes

### PR DESCRIPTION
**Pull request by CharlesRendle**
PR addressing issue #651

Adds unit tests for all API functions and cached properties which have been implemented.

Moves former test areas covered by behave tests to unit tests

We've added a helper function to store `csv_rdf_managers` and `state objects` for commonly used test-case datasets to improve the performance of our unit tests.

Based of the test times on Github and locally the test times for unit test have been halved.

### CODE REVIEW

**Comment by CharlesRendle** [comment]
@nimshi89 for tomorrow - you may want to look into using the cached state object functions in `test_data_cube_state.py` instead of having variables such as `data_cube_state3`. This is the same in `test_csvw_state.py`.

I also forgot when we changed the name of the cache helper, we probably should have changed the function names to reflect that they will be called `inspectors` in the future

File: tests/unit/cli/inspect/test_metadataprinter.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+from tests.unit.cli.inspect.test_inspectdatasetmanager import (
+    expected_dataframe_pivoted_multi_measure,
+    expected_dataframe_pivoted_single_measure,
+)
+from tests.unit.test_baseunit import get_test_cases_dir
+
+_test_case_base_dir = get_test_cases_dir() / "cli" / "inspect"
+
+
+def test_multi_measure_pivoted_shape_observation_info():
```
Comment (robons) [inline tests/unit/cli/inspect/test_metadataprinter.py:23]: We're also testing the value count information as well. Could you include that in the name/description or else split this unit test into one which tests the obs info and the other which tests the value counts?

Could you also make clear in the description that this is testing a cube (as opposed to a code list).

File: tests/unit/utils/sparqlhandler/test_data_cube_state.py [hunk 2]
```diff
@@ -155,19 +161,26 @@ def test_get_dsd_qube_components_for_csv():
...
         csvw_rdf_manager.csvw_state.get_primary_catalog_metadata()
     )
 
     data_set_uri = primary_catalog_metadata.dataset_uri
-    data_set_uri = to_absolute_rdflib_file_path(data_set_uri, csvw_metadata_json_path)
+    data_set_uri = to_absolute_rdflib_file_path(
+        data_set_uri,
+        csvw_metadata_json_path,
+    )
     csv_url = select_qb_csv_url(
```
Comment (robons) [inline tests/unit/utils/sparqlhandler/test_data_cube_state.py:184]: Can you replace this sparql query call here with a call to the `data_cube_state.get_cube_identifiers_for_data_set(...)` function?

File: tests/unit/utils/sparqlhandler/test_data_cube_state.py [hunk 3]
```diff
@@ -126,19 +125,26 @@ def test_get_data_set_dsd_csv_url_for_csv_url():
...
```
Comment (robons) [inline tests/unit/utils/sparqlhandler/test_data_cube_state.py:148]: Can you replace this sparql query call here with a call to the `data_cube_state.get_cube_identifiers_for_data_set(...)` function?

File: tests/unit/utils/sparqlhandler/test_data_cube_state.py [hunk 4]
```diff
@@ -268,14 +281,257 @@ def test_detect_csvw_shape_standard():
...
+
+    assert cube_identifiers is not None
+    assert cube_identifiers.csv_url == "energy-trends-uk-total-energy.csv"
+    assert cube_identifiers.data_set_url == "energy-trends-uk-total-energy.csv#dataset"
+    assert cube_identifiers.dsd_uri == "energy-trends-uk-total-energy.csv#structure"
+
+
+def test_get_cube_identifiers_for_data_set_error():
+    """
+    Ensures we can return cube identifiers from a given dataset_uri
```
Comment (robons) [inline tests/unit/utils/sparqlhandler/test_data_cube_state.py:323]: This description isn't quite right.

File: tests/unit/utils/sparqlhandler/test_data_cube_state.py [hunk 5]
```diff
@@ -268,14 +281,257 @@ def test_detect_csvw_shape_standard():
...
+        ComponentPropertyType.Measure,
+        "Some Other Measure",
+        ["Some Other Obs Val"],
+        ["Some Other Obs Val"],
+        "qb-id-10003.csv#structure",
+        True,
+    )
+
+
+def test_dsd_single_measure_pivoted_shape():
```
Comment (robons) [inline tests/unit/utils/sparqlhandler/test_data_cube_state.py:448]: Do we have any tests for the DSD components of the standard shape?

**Review by robons — COMMENTED** [review]
Looks pretty good. I've left a few questions and comments.

**Comment by SarahJohnsonONS** [comment]
Nick and I had a look - looks great, and we particularly liked the cached inspector helpers - will make writing new tests much simpler.

File: tests/unit/utils/sparqlhandler/test_data_cube_state.py (hunk 2 above)
Comment (CharlesRendle) [reply tests/unit/utils/sparqlhandler/test_data_cube_state.py]: Replaced sparql query call with access via `data_cube_state` and removed associated imports.

File: tests/unit/utils/sparqlhandler/test_data_cube_state.py (hunk 3 above)
Comment (CharlesRendle) [reply tests/unit/utils/sparqlhandler/test_data_cube_state.py]: Replaced sparql query call with access via `data_cube_state` and removed associated imports.

File: tests/unit/utils/sparqlhandler/test_data_cube_state.py (hunk 4 above)
Comment (CharlesRendle) [reply tests/unit/utils/sparqlhandler/test_data_cube_state.py]: Altered description

File: tests/unit/utils/sparqlhandler/test_data_cube_state.py (hunk 5 above)
Comment (CharlesRendle) [reply tests/unit/utils/sparqlhandler/test_data_cube_state.py:437]: Added a unit test for the DSD components of the standard shape single unit single measure dataset: `/ "single-unit_single-measure" / "energy-trends-uk-total-energy.csv-metadata.json"`

**Comment by CharlesRendle** [comment]
Have [created a ticket](https://app.zenhub.com/workspaces/features-squad-61b72275ac896c0010a1b00b/issues/gh/gss-cogs/csvcubed/728) for implementing the `inspectors_cache` functions in the new unit tests written in subsequent tickets

File: tests/unit/cli/inspect/test_metadataprinter.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+            },
+            {
+                "Measure": "Some Other Measure",
+                "Unit": "Percent",
+                "Count": 3,
+            },
+        ]
+    )
+    assert result_dataset_value_counts.by_measure_and_unit_val_counts_df.empty == False
+    assert type(expected_df.columns[0]) == type(
```
Comment (robons) [inline tests/unit/cli/inspect/test_metadataprinter.py:81]: Any idea why we're asserting these types are equivalent here? Surely it's just going to be saying `pd.Series == pd.Series` or similar?

**Review by robons — APPROVED** [review]

File: tests/unit/cli/inspect/test_metadataprinter.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+    data_cube_state = get_data_cube_inspector(path_to_json_file)
+    metadata_printer = MetadataPrinter(data_cube_state)
+
+    result_code_lists: CodelistsResult = metadata_printer.result_code_lists
+    assert result_code_lists is not None
+
+    assert len(result_code_lists.codelists) == 1
+    assert (
+        first(result_code_lists.codelists, lambda c: c.cols_used_in == "Some Dimension")
```
[the remark this answers is not in the document]
Comment (CharlesRendle) [reply tests/unit/cli/inspect/test_metadataprinter.py:40]: Changed.

File: tests/unit/cli/inspect/test_metadataprinter.py (hunk 6 above)
Comment (CharlesRendle) [reply tests/unit/cli/inspect/test_metadataprinter.py:81]: This was me while debugging. Not needed anymore. Removed.

**Pull request merged**